File: solution/1653. Minimum Deletions to Make String Balanced/1653_02.cpp
class Solution {
public:
    int minimumDeletions(string s) {
        int n = s.size();
        s = '#' + s;
        int dpi = 0, bCount = 0;
        for (int i = 1; i <= n; ++i) {
            if (s[i] == 'a') {
                // keep 'a' => bCount
                // delete 'a' => dp[i - 1] + 1
                // dp[i] = min(bCount, dp[i - 1] + 1);
                dpi = min(bCount, dpi + 1);
            } else {
                // dp[i] = dp[i - 1];
                bCount++;
            }
        }
        return dpi;
    }
};

// dp[i] := a...a 或者 a...ab...b 情況下的 Minimum Deletions

// aababbab
//   i

File: solution/1653. Minimum Deletions to Make String Balanced/1653_02.py
class Solution:
    def minimumDeletions(self, s: str) -> int:
        n = len(s)
        s = '#' + s
        dpi = b_count = 0
        for i in range(1, n + 1):
            if s[i] == 'a':
                dpi = min(dpi + 1, b_count)
            else:
                b_count += 1
        return dpi
    
